packages/internal-test-utils/ReactInternalTestUtils.js

```diff
@@ -264,3 +264,40 @@ ${diff(expectedLog, actualLog)}
   Error.captureStackTrace(error, assertLog);
   throw error;
 }
+
+// Simulates dispatching events, waiting for microtasks in between.
+// This matches the browser behavior, which will flush microtasks
+// between each event handler.
+export async function dispatchAndWaitForDiscrete(
+  node: Node,
+  eventType: string,
+): Promise<void> {
+  assertYieldsWereCleared(waitForDiscrete);
+
+  // Bubbling phase
+  // Walk the path from the element to the document and dispatch
+  // the event on each node, flushing microtasks in between.
+  for (let current = node; current; current = current.parentNode) {
```

Comment:
Technically if you remove or move a node during this phase it'll get the events.

So really this should create a snapshot of all parent nodes into an array and then dispatch to each of those. That way if they change the DOM during the bubbling phase, all events still happen (on disconnected nodes).

Reply:
Also, if this doesn't end up with a parentNode that is in a document, it shouldn't dispatch anything because it means the target was detached before dispatching.

Reply:
Fixed and added a test.